packages/grafana-data/src/types/data.ts

```diff
@@ -213,3 +213,27 @@ export interface DataConfigSource {
 
 type Truthy<T> = T extends false | '' | 0 | null | undefined ? never : T;
 export const isTruthy = <T>(value: T): value is Truthy<T> => Boolean(value);
+
+/**
+ * Serves no runtime purpose - only used to make typescript check a value has been correctly
+ * narrowed to an object
+ */
+function identityObject(value: object): object {
+  return value;
+}
```

Comment:
Is the jsdoc comment on both `identityObject` and `isObject` not clear enough?

User-defined type predicates are not type safe (it can return whatever it want, the narrowing is not type checked), and thus can easily be a source of introducing bugs and unpredictability.

`identityObject` is a function that typescript knows should only accept an argument of type `object`. Calling this in `isObject` is a build-time (not runtime) test that we have correctly narrowed `value` from `unknown` to `object`.

For example, if a developer were to come along and remove the `value !== null` check from `isObject`, they would immediately see a typescript error because now `object | null` would be passed into `identityObject`.

In other words, it serves no runtime purpose - only used to make typescript check a value has been correctly narrowed to an object.

Reply:
Yeah, it makes sense. Could we fulfil the same thing with a couple of tests?